rust-symcrypt/src/hmac.rs

```diff
@@ -1,180 +1,419 @@
+                    symcrypt_sys::SymCryptHmacSha256Init(
+                        &mut instance.inner.state,
+                        &*instance.inner.expanded_key.0,
+                    );
+                    Ok(instance)
+                }
+                err => Err(err.into()),
+            }
         }
-        instance
     }
+}
 
-    pub fn append(&mut self, data: &[u8]) {
+impl HmacState for HmacSha256State {
+    type Result = [u8; SHA256_HMAC_RESULT_SIZE];
+
+    fn append(&mut self, data: &[u8]) {
         unsafe {
+            // SAFETY: FFI calls
             symcrypt_sys::SymCryptHmacSha256Append(
-                &mut self.state,
+                &mut self.inner.state,
                 data.as_ptr(),
                 data.len() as symcrypt_sys::SIZE_T,
             )
         }
     }
 
-    pub fn result(&mut self, result: &mut [u8; SHA256_HMAC_RESULT_SIZE]) {
+    fn result(mut self) -> Self::Result {
+        let mut result = [0u8; SHA256_HMAC_RESULT_SIZE];
         unsafe {
-            symcrypt_sys::SymCryptHmacSha256Result(&mut self.state, result.as_mut_ptr());
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptHmacSha256Result(&mut self.inner.state, result.as_mut_ptr());
+        }
+        result
+    }
+
+    fn copy(&self) -> Box<Self> {
+        let mut new_state = HmacSha256State {
+            inner: Box::pin(HmacSha256Inner {
+                state: symcrypt_sys::SYMCRYPT_HMAC_SHA256_STATE::default(),
+                expanded_key: Arc::clone(&self.inner.expanded_key),
+            }),
+        };
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptHmacSha256StateCopy(
+                &self.inner.state,
+                &*self.inner.expanded_key.0,
+                &mut new_state.inner.state,
+            );
         }
+        Box::new(new_state)
     }
 }
 
 impl Drop for HmacSha256State {
     fn drop(&mut self) {
         unsafe {
+            // SAFETY: FFI calls
             symcrypt_sys::SymCryptWipe(
-                ptr::addr_of_mut!(self.state) as *mut c_void,
-                mem::size_of_val(&mut self.state) as symcrypt_sys::SIZE_T,
-            );
-
-            symcrypt_sys::SymCryptWipe(
-                ptr::addr_of_mut!(self.expanded_key) as *mut c_void,
-                mem::size_of_val(&mut self.expanded_key) as symcrypt_sys::SIZE_T,
+                ptr::addr_of_mut!(self.inner.state) as *mut c_void,
+                mem::size_of_val(&mut self.inner.state) as symcrypt_sys::SIZE_T,
             );
         }
     }
 }
 
-pub fn hmac_sha256(key: &[u8], data: &[u8], result: &mut [u8; SHA256_HMAC_RESULT_SIZE]) {
+/// Stateless Hmac functions for Sha256
+pub fn hmac_sha256(
+    key: &[u8],
+    data: &[u8],
+    result: &mut [u8; SHA256_HMAC_RESULT_SIZE],
+) -> Result<(), SymCryptError> {
     unsafe {
+        // SAFETY: FFI calls
         let mut expanded_key = symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY::default();
```

Comment:
>  let mut expanded_key [](http://example.com/codeflow?start=6&length=22)

You want to wrap this via HmacSha256ExpandedKey() so it naturally gets dropped and wiped.